src/actor/npc/ac_npc_shop_master.c

```diff
@@ -0,0 +1,2677 @@
+    (*Common_Get(clip).npc_clip->init_proc)(actorx, game);
+}
+
+static void aNSM_actor_draw(ACTOR* actorx, GAME* game) {
+    (*Common_Get(clip).npc_clip->draw_proc)(actorx, game);
+}
+
+static int aNSM_get_zone(xyz_t *wpos) {
+    if (wpos->x < (f32)120.0) {
+        if (wpos->z < (f32)180.0) {
+            return 1;
+        } else {
+            return 2;
+        }
+    }
+    if (wpos->z < (f32)160.0) {
+        return 0;
+    } else {
+        return 3;
+    }
+}
+
+static u8 aNSM_get_next_zone(u8 p1, u8 p2) {
+    static u8 next_zone[16] = {0, 1, 1, 1, 0, 1, 2, 2, 1, 1, 2, 3, 2, 2, 2, 3};
+    int t = p2 * 4;
+    return next_zone[t + p1];
+}
+
+static void aNSM_search_player(NPC_SHOP_MASTER_ACTOR *shop_master, GAME_PLAY *play) {
+    ACTOR *actorx = (ACTOR *)shop_master;
+    PLAYER_ACTOR *player = get_player_actor_withoutCheck(play); 
+
+    if (player != NULL) {
+        s16 angle = actorx->shape_info.rotation.y - shop_master->player_angle;
+        if (ABS(angle) > 0x4000) {
+            aNSC_setupAction(shop_master, play, 0x42);
+        } else {
+            chase_angle(&actorx->shape_info.rotation.y, shop_master->player_angle, 0x800);
+            actorx->world.angle.y = actorx->shape_info.rotation.y;
+        }
+    }
+}
+
+static void aNSM_search_player2(NPC_SHOP_MASTER_ACTOR *shop_master, GAME_PLAY *play) {
+    static f32 posX[4] = { 220.0, 100.0, 100.0, 220.0 };
+    static f32 posZ[4] = { 140.0, 140.0, 220.0, 220.0 };
+    PLAYER_ACTOR *player = get_player_actor_withoutCheck(play);
+    if (player != NULL) {
+            f32 dz = posX[shop_master->_9A3] - shop_master->npc_class.actor_class.world.position.x;
+            f32 dx = posZ[shop_master->_9A3] - shop_master->npc_class.actor_class.world.position.z;
+
+            s16 angle = atans_table(dx, dz);
+            chase_angle(&shop_master->npc_class.actor_class.shape_info.rotation.y, angle, 0x800);
+
+            shop_master->npc_class.actor_class.world.angle.y = shop_master->npc_class.actor_class.shape_info.rotation.y;
+
+            if (dz * dz + dx * dx < (f32)0.0) {
+                shop_master->_9A3 = aNSM_get_next_zone(shop_master->_9A4, shop_master->_9A2);
+            }
+    }
+
+}
+
+static int aNSM_check_safe_zone(NPC_SHOP_MASTER_ACTOR *shop_master, PLAYER_ACTOR *player) {
+    int res = FALSE;
+    if (player->actor_class.world.position.z > (f32)280.0 && shop_master->_9A2 == 3) {
+        res = TRUE;
+    }
+    return res;
+}
+
+static void aNSM_actor_move(ACTOR *actorx, GAME *game) {
+    NPC_SHOP_MASTER_ACTOR *shop_master = (NPC_SHOP_MASTER_ACTOR *)actorx;
+    GAME_PLAY *play = (GAME_PLAY *)game;
+    PLAYER_ACTOR *player = get_player_actor_withoutCheck(play);
+    CLIP(npc_clip)->move_before_proc(actorx, game);
+    aNSC_BGcheck(actorx);
+    aNSC_set_zone_data(shop_master, (ACTOR *)player);
+    aNSC_set_player_angl(shop_master);
+    ((void (*)(ACTOR *actorx, GAME *game))(shop_master->proc))(actorx, game);
```

Comment:
Whatever is happening here has to be cleaned up. If you have to cast a function pointer like this then there's likely an issue with NPC_SHOP_MASTER_ACTOR::proc's function pointer type.